cpp/server/certificate_handler.cc

```diff
@@ -0,0 +1,181 @@
+#include <functional>
+
+#include "log/frontend.h"
+#include "server/certificate_handler.h"
+#include "server/json_output.h"
+#include "util/json_wrapper.h"
+#include "util/status.h"
+#include "util/thread_pool.h"
+
+namespace cert_trans {
+
+using ct::LogEntry;
+using ct::SignedCertificateTimestamp;
+using std::bind;
+using std::make_shared;
+using std::multimap;
+using std::placeholders::_1;
+using std::shared_ptr;
+using std::string;
+using std::unique_ptr;
+using util::Status;
+
+
+namespace {
+
+
+bool ExtractChain(JsonOutput* output, evhttp_request* req, CertChain* chain) {
+  if (evhttp_request_get_command(req) != EVHTTP_REQ_POST) {
+    output->SendError(req, HTTP_BADMETHOD, "Method not allowed.");
+    return false;
+  }
+
+  // TODO(pphaneuf): Should we check that Content-Type says
+  // "application/json", as recommended by RFC4627?
+  JsonObject json_body(evhttp_request_get_input_buffer(req));
+  if (!json_body.Ok() || !json_body.IsType(json_type_object)) {
+    output->SendError(req, HTTP_BADREQUEST, "Unable to parse provided JSON.");
+    return false;
+  }
+
+  JsonArray json_chain(json_body, "chain");
+  if (!json_chain.Ok()) {
+    output->SendError(req, HTTP_BADREQUEST, "Unable to parse provided JSON.");
+    return false;
+  }
+
+  VLOG(2) << "ExtractChain chain:\n" << json_chain.DebugString();
+
+  for (int i = 0; i < json_chain.Length(); ++i) {
+    JsonString json_cert(json_chain, i);
+    if (!json_cert.Ok()) {
+      output->SendError(req, HTTP_BADREQUEST,
+                        "Unable to parse provided JSON.");
+      return false;
+    }
+
+    unique_ptr<Cert> cert(new Cert);
+    cert->LoadFromDerString(json_cert.FromBase64());
+    if (!cert->IsLoaded()) {
+      output->SendError(req, HTTP_BADREQUEST,
+                        "Unable to parse provided chain.");
+      return false;
+    }
+
+    chain->AddCert(cert.release());
+  }
+
+  return true;
+}
+
+
+}  // namespace
+
+
+CertificateHttpHandler::CertificateHttpHandler(
+    LogLookup<LoggedEntry>* log_lookup,
+    const ReadOnlyDatabase<LoggedEntry>* db,
+    const ClusterStateController<LoggedEntry>* controller,
+    const CertChecker* cert_checker, Frontend* frontend, ThreadPool* pool,
+    libevent::Base* event_base)
+    : HttpHandler(log_lookup, db, controller, pool, event_base),
+      cert_checker_(cert_checker),
```

Comment:
`HttpHandler` checks all of its params for not-nullness, `frontend` is permitted to be null, but I've added a check for `cert_checker`